connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoClient.scala

```diff
@@ -1,59 +1,82 @@
 package com.microsoft.kusto.spark.utils
 
-import com.microsoft.azure.kusto.data.{Client, ClientFactory, ConnectionStringBuilder}
-import com.microsoft.azure.kusto.ingest.{IngestClient, IngestClientFactory}
-import com.microsoft.kusto.spark.datasource._
-import com.microsoft.kusto.spark.utils.{KustoDataSourceUtils => KDSU, KustoConstants => KCONST}
-
-import scala.collection.immutable.HashMap
-
-object KustoClient {
-  var ingestClientCache = new HashMap[String, IngestClient]
-  var adminClientCache = new HashMap[String, Client]
-  org.apache.spark.sql.sources.In
-  def getAdmin(authentication: KustoAuthentication, clusterAlias: String, isIngestCluster: Boolean = false): Client = {
-    val clusterUri = s"https://${if(isIngestCluster) "ingest" else ""}$clusterAlias.kusto.windows.net"
-    val kcsb = getKcsb(authentication,clusterUri)
-    kcsb.setClientVersionForTracing(KCONST.clientName)
-    getAdmin(kcsb, clusterAlias, isIngestCluster)
-  }
 
-  def getAdmin(kcsb: ConnectionStringBuilder, clusterAlias: String, isIngestCluster: Boolean): Client = {
-    var cachedClient: Option[Client] = adminClientCache.get(clusterAlias + isIngestCluster)
-    if(cachedClient.isEmpty){
-      cachedClient = Some(ClientFactory.createClient(kcsb))
-      adminClientCache = adminClientCache + (clusterAlias + isIngestCluster -> cachedClient.get)
-    }
+import java.util.StringJoiner
 
-    cachedClient.get
-  }
+import com.microsoft.azure.kusto.data.Client
+import com.microsoft.azure.kusto.ingest.IngestClient
+import com.microsoft.kusto.spark.datasource.KustoCoordinates
+import com.microsoft.kusto.spark.datasource.KustoOptions.SinkTableCreationMode
+import com.microsoft.kusto.spark.datasource.KustoOptions.SinkTableCreationMode.SinkTableCreationMode
+import com.microsoft.kusto.spark.utils.CslCommandsGenerator._
+import com.microsoft.kusto.spark.utils.KustoDataSourceUtils.extractSchemaFromResultTable
+import com.microsoft.kusto.spark.utils.{KustoDataSourceUtils => KDSU}
+import org.apache.spark.sql.types.StructType
+import org.joda.time.{DateTime, DateTimeZone, Period}
 
-  def getIngest(authentication: KustoAuthentication, clusterAlias: String): IngestClient = {
-    val ingestKcsb = getKcsb(authentication, s"https://ingest-$clusterAlias.kusto.windows.net")
-    ingestKcsb.setClientVersionForTracing(KCONST.clientName)
-    getIngest(ingestKcsb, clusterAlias)
-  }
+import scala.collection.JavaConverters._
+
+class KustoClient(val clusterAlias: String, val engineClient: Client, val dmClient: Client, val ingestClient: IngestClient) {
```

Comment:
I would avoid calling this class `KustoClient` to prevent ambiguity. 
A better name would be `KustoBackend` IMHO, or just `KustoDatabase` as it has some database related logic

Reply:
We'll think about it. Same as above.